first round of refactors

mhario opened the pull request:
Working on fleshing out the DB. This PR is rough, just opened to view differences and chat! 

I think the biggest refactor that I would like to make would be to pull out 'contact info' into its own table and then link to it from other tables--which would simplify things and reduce duplicate data. In the process, I think we could also break up 'users' into separate tables: one for 'advocates' and one for 'teams' for instance. As it is right now I have multiple enums on the 'user' table and I think its working too hard! 

This current PR is meant to look very similar to the google doc in the dev folder.

mhario committed:
first round of refactors

bethqiang commented on db/models/ally.js:
```diff
@@ -0,0 +1,29 @@
+
+const Ally = db.define('allies', {
+  username: Sequelize.STRING,
+  phone_number: {
```
Is having their #s absolutely necessary? Is there reasoning behind this?

mhario replied:
I was thinking it would be meaningful to require a phone number of the advocates so that they can be available for our clients. That way a woman who comes to us asking for help always has a personal phone number they can call. That was the intention, what do you think?

bethqiang replied:
I like the idea, but at least personally, I'm super wary about giving out my phone number, regardless of whom it's to...maybe something for us to all chat about. If anything, I'd maybe personally vote for when we're onboarding/training, just say we'd love it if you felt comfortable giving your number to your women, but it's not required?

Also, having to ALWAYS be available to someone doesn't sound like the best idea either; people lead busy and important lives...

mhario replied:
True, I just wanted to make sure people were available as needed. But I Guess that concern is better-addressed by the 'urgent' list anyway. I will take this off

bethqiang commented on db/models/ally.js:
```diff
@@ -0,0 +1,29 @@
+  job_title: Sequelize.STRING,
+  company: Sequelize.STRING,
+  industry: Sequelize.ARRAY(Sequelize.STRING),
+  max_clients: Sequelize.INTEGER
```
Is this referring to the number of cases they can take on?

mhario replied:
yeah, should I change the name? this was a variable listed on the google doc

bethqiang replied:
Oh, is this something they input when they sign up saying "I'm willing to do x number"?

(That would make a lot more sense. I was for some reason thinking we were imposing a limit on them?)

bethqiang commented on db/models/ally.js:
```diff
@@ -0,0 +1,29 @@
+  max_clients: Sequelize.INTEGER
+}, {
+  indexes: [
+    { unique: false, fields: ['industry'] }
```
Why are we indexing by industry?

mhario replied:
I think we were planning on trying to match women with advocates from a similar industry so this presents an easy way to do that. It may be unnecessary, or maybe there is a better way?

bethqiang replied:
I thought it was just weird that this was one of the few indexes I saw...I'd think we'd index email and/or sign-in info before this. Something else to chat about potentially?

bethqiang commented on db/models/donation.js:
```diff
@@ -0,0 +1,27 @@
+    defaultValue: Sequelize.NOW
+  },
+  amount: {
+    type: Sequelize.INTEGER, // in U.S. Cents
```
Appreciate the comment! 👍

mhario replied:
:)

bethqiang commented on db/models/donation.js:
```diff
@@ -0,0 +1,27 @@
+    defaultValue: false
+  },
+  dedication: {
+    type: Sequelize.STRING
```
Are they going to be inputting something like "twitter - @bigdonorhere"?

Also, random big dream I just had: if we ever do send out newsletters to like not just our members (like whomever from the general public signs up to receive them because like...news or something idk), we can give the option of donors being thanked in that too?

mhario replied:
I was thinking the dedication would be a pretty open-ended message string.  I was thinking we could have a list of donors with their custom messages. Although I'm realizing now that I don't have a field for the actual donor name lol

bethqiang replied:
Donor name, and in that case, the way in which they want to be thanked (if they do at all) (e.g. Twitter, etc.).

mhario replied:
That 'thanked' flag was something Rachel wanted, I wasn't sure if she had thought through how she wanted this to work or not. I think the thank-you could really be given on the page following the donation, but I think she wanted to be more personal

bethqiang replied:
Oh I just meant - though we have a thanked flag - we wouldn't know in what manner they'd want to be thanked (whether over Twitter, or like a community Slack channel, or in a non-existent email newsletter to all of our adoring fans, etc.)

mhario replied:
ohh I see, I misunderstood.. yeah you're right though!

bethqiang commented on db/models/feedback.js:
```diff
@@ -0,0 +1,18 @@
+const Sequelize = require('sequelize')
+const db = require('APP/db')
+
+const Feedback = db.define('feedback', {
```
Feedback for...what, exactly? (And does there need to be a "feedback for" field?)

mhario replied:
So this was a small adjustment I made to the google doc. We have a spreadsheet with all of our volunteers, but i think the same form offered the option to provide anonymous/ random feedback. I thought it was odd that both data were on the same table so I broke one out here

bethqiang replied:
So...this is feedback from...whom? And to whom?

mhario replied:
From people, for us. In the google doc there are already one or two points of feedback

bethqiang commented on db/models/feedback.js:
```diff
@@ -0,0 +1,18 @@
+    defaultValue: Sequelize.NOW
+  },
+  // e-mail not validated here because email is not required
+  // email should be checked for validity on front-end
```
Why not also have Sequelize check it too? Always good to have a backup. :woman_shrugging:

mhario replied:
I did a bit of looking into it but I didn't find an easy way to check that an email was a valid email, without requiring the field--ideally it would be a valid email, or null. I'm sure there's a better way to do this though

bethqiang replied:
```
validate: {
  isEmail: true
}
```

See lines 15-17 of the Ally model.

mhario replied:
Oops, yeah I was thinking that that would require the email field not be null, also.

bethqiang commented on db/models/user.js:
```diff
@@ -8,15 +8,26 @@ const User = db.define('users', {
   username: Sequelize.STRING,
   first_name: Sequelize.STRING,
   last_name: Sequelize.STRING,
+  phone_number: Sequelize.STRING,
```
Again, I'm also sort of personally iffy as to whether or not this is necessary...and for all of admin stuff, we can put that somewhere else?

bethqiang commented on db/models/user.js:
```diff
@@ -8,15 +8,26 @@ const User = db.define('users', {
+  },
+  team: {
+    type: Sequelize.ENUM('Dev', 'Social Media', 'Fundraising', 'Crowdsurfing'),
+    allowNull: false
```
But what if I'm an advocate and I don't belong to a team? (I know you want to make them a team, but just saying, the way this is set up now...)

mhario replied:
right, this isn't meant to be totally complete. I was figuring there would be more teams to add as well.

bethqiang commented on db/models/volunteer.js:
```diff
@@ -4,7 +4,7 @@ const Sequelize = require('sequelize')
 const db = require('APP/db')
 
 const Volunteer = db.define('volunteers', {
```
Are we not asking for email or any kind of contact info? Is this separate from allies/advocates? If so, then what is a volunteer? (Is this only for when they initially contact WWA, then once they're reviewed they get transferred into the ally/advocate table?)

mhario replied:
I think that is what it was. I really think we could benefit from a bit of a vocabulary clarification honestly. For me, the volunteers are like you described: people who 'just signed up' and are awaiting assignment to a team, or validation, or whatever.

bethqiang replied:
:+1: to vocab clarification. Maybe a separate Google doc for that?

bethqiang commented:
In terms of a higher level overview - when you say pull out 'contact info', whose contact info are you referring to? Everyone's? Just one group?

I like the idea of breaking users up - one for the women, one for people who are working for WWA (admins, devs, social media, fundraising, other vols), one for advocates, one for allies maybe? All have a user ID, and then if there's someone that meets more than one critera...oof. Does that mean we have to have a huge table too? How would that join work? (I do like the idea. Just think we might need to think about it a little more.)

mhario commented:
So my thinking as far as overall architecture would be that we could have a 'contact info' table which has phone number / email etc, for every person we know. We could use hooks to require names of everyone EXCEPT women who are asking for help ('clients'?) 

Anyone who is on two different teams could simply be on two different tables. The only redundant data would be the userid column on their second team, right? I think that would result in much simpler 'team' tables, instead of having one huge 'user' table. We could also simply have a 'no-team-yet' table, for volunteers who still need to be assigned, or who won't be on one at all. 

Does that make sense?  I guess their userid would be on the 'contacts' table which is where a new user is first registered. The team tables would then be a list of those userids

mhario commented on db/models/ally.js:
```diff
@@ -0,0 +1,29 @@
+  job_title: Sequelize.STRING,
+  company: Sequelize.STRING,
+  industry: Sequelize.ARRAY(Sequelize.STRING),
+  max_clients: Sequelize.INTEGER
```
👍

bethqiang commented:
So, wait, would the contact info have just team/volunteers/allies/advocates then?

So it'd generally look something like--

Contact info: phone number, email, name, user_id (<< I personally think we should use user_id as the foreign key, but I also like the idea of having a name in the table...urgh)...for...everyone?
Teams: user_id, name, team: enum['dev', 'fundraising', 'etc']
(Or are you saying each team is its own table?)

We should def meet about this and hammer all of this out + details. Would probs be easier than trying to do this through GitHub, haha.

mhario commented:
Yeah couldn't agree more about meeting up and talking this all through haha Maybe even faster to just pair program the DB. I can probably meet with you today if you like. I can't remember if you were busy today or yesterday

mhario commented:
We also need to add the notes-for-cases functionality

bethqiang committed:
Merge branch 'master' into database_refactor

bethqiang committed:
Add discussed fields to user (staff) model

bethqiang committed:
Add allowNulls & related name tests

- allowNull to user models
- Add beforeEach and afterEach in tests
- Add tests for name cannot be null & cannot be empty
- Edit authentication tests to work with beforeEach/afterEach
